# 2.1 Regular Expressions

The book aptly calls out regular expressions as one of the unsung successes in computer science :-). They are super handy for a lot of tasks and can sometimes be adequate to solve some NLP problems. They can also help establish baselines for many NLP problems.

In Python, `re` is the default regular expression library that we need to import and use.

In [1]:
import re

Below, we will use the variable `s` for the string that we are interested in searching in and a variable ending in `_re` to denote the regex we are searching for. We will mostly use the function `re.search` since it's the most flexible. We will also use compiled regexes since they run faster.

## 2.1.1 Basic Regular Expression Patterns

In [2]:
s = '''Do you know what a woodchuck is?'''

# compile a regex for matching the exact string 'woodchuck'
# we use raw strings so regex special characters are not interpreted
woodchuck_re = re.compile(r'woodchuck')

# search
m = re.search(woodchuck_re, s)
print(m)

<re.Match object; span=(19, 28), match='woodchuck'>


In [3]:
s[19:28]

'woodchuck'

In [4]:
# regexes are case-sensitive by default.

s = '''I'm called little Buttercup'''

my_re = re.compile(r'buttercup')

# this won't match since the case does not match.
m = re.search(my_re, s)
print(m)

None


In [5]:
# use square braces for disjunction to allow case insensitivity 
# for the first character

my_re = re.compile(r'[bB]uttercup')

m = re.search(my_re, s)
print(m)

<re.Match object; span=(18, 27), match='Buttercup'>


In [6]:
# match any vowel, case-insensitive

s = '''Hello there, Mr. E!'''

vowel_re = re.compile(r'[aeiouAEIOU]')

m = re.search(vowel_re, s)
print(m)

<re.Match object; span=(1, 2), match='e'>


Notice that only the first match is returned by `re.search`. If we want to find all the matches, we can use `re.finditer`

In [7]:
for m in re.finditer(vowel_re, s):
    print(m)

<re.Match object; span=(1, 2), match='e'>
<re.Match object; span=(4, 5), match='o'>
<re.Match object; span=(8, 9), match='e'>
<re.Match object; span=(10, 11), match='e'>
<re.Match object; span=(17, 18), match='E'>


In [8]:
# Set case-insensitive match with flag

s = '''Hello there, Mr. E!'''

vowel_re = re.compile(r'[aeiou]', flags=re.IGNORECASE)

m = re.search(vowel_re, s)
print(m)

<re.Match object; span=(1, 2), match='e'>


In [9]:
# match any uppercase letter

s = '''hello there, Mr. E!'''

upper_re = re.compile(r'[A-Z]')
m = re.search(upper_re, s)
print(m)

<re.Match object; span=(13, 14), match='M'>


Similarly, `[a-z]` matches any lower-case letter and `[0-9]` matches any digit. In python `\d` also means a digit.

In [10]:
s = '''He won $100 in the raffle.'''

# let's try a regex directly and not compile it
for m in re.finditer(r'\d', s):
    print(m)

<re.Match object; span=(8, 9), match='1'>
<re.Match object; span=(9, 10), match='0'>
<re.Match object; span=(10, 11), match='0'>


Negation i.e. to not match on something, use `[^`.

In [11]:
s = '''Oyfn PripeT'''

not_upper_case_re = re.compile(r'[^A-Z]')

for m in re.finditer(not_upper_case_re, s):
    print(m)

<re.Match object; span=(1, 2), match='y'>
<re.Match object; span=(2, 3), match='f'>
<re.Match object; span=(3, 4), match='n'>
<re.Match object; span=(4, 5), match=' '>
<re.Match object; span=(6, 7), match='r'>
<re.Match object; span=(7, 8), match='i'>
<re.Match object; span=(8, 9), match='p'>
<re.Match object; span=(9, 10), match='e'>


For substituting patterns matching a regular expression with something else, we can use `re.sub`.

In [12]:
# remove all upper-case characters

s = '''Hello There, Doc!'''

upper_removed_s = re.sub(r'[A-Z]', '', s)
print(upper_removed_s)

ello here, oc!


In [13]:
# optional element, meaning 0 or 1 occurrence

# what they see in England
s = 'rainbow colours'

m = re.search(r'colou?r', s)
print(m)

<re.Match object; span=(8, 14), match='colour'>


In [14]:
# what they see in the US
s = 'rainbow colors'

m = re.search(r'colou?r', s)
print(m)

<re.Match object; span=(8, 13), match='color'>


### Kleene * (cleany star) and Kleene +

When we want to match zero or more occurrences of the immediately previous character or regex. So `a*` means 0 or more a's. It is greedy and will try to mach for maximum length first.

In [15]:
s = 'aaaabracadabraa'
re.search(r'a*', s)

<re.Match object; span=(0, 4), match='aaaa'>

In [16]:
s = 'freezing temps'
re.search(r'a*', s)

<re.Match object; span=(0, 0), match=''>

But the regex also matches the beginning of the string, which is an empty character. If we want to match 1 or more occurrences, then we need to use "Kleene +"

In [17]:
s = 'freezing temps'
m = re.search(r'a+', s)
print(m)

None


In [18]:
# find all runs of 1 or more a's in a string
s = 'aabracaaada braaaaaa..'
for m in re.finditer(r'a+', s):
    print(m)

<re.Match object; span=(0, 2), match='aa'>
<re.Match object; span=(4, 5), match='a'>
<re.Match object; span=(6, 9), match='aaa'>
<re.Match object; span=(10, 11), match='a'>
<re.Match object; span=(14, 20), match='aaaaaa'>


---

Sheep langugage consists of all strings starting with a 'b', followed by at least two 'a's, followed by an exclamation point, so like:

- baa!
- baaa!
- baaaa!
- baaaaa!
- ...

In [19]:
s = '''
Sheep 1 said, "baaaaaa!".
Sheep 2 responded, "baaa!".
The rest of the sheep said, "baaaaaaaaaa!".
'''

In [20]:
for m in re.finditer(r'baa+!', s):
    print(m)

<re.Match object; span=(16, 24), match='baaaaaa!'>
<re.Match object; span=(47, 52), match='baaa!'>
<re.Match object; span=(84, 96), match='baaaaaaaaaa!'>


We can also use the regex `ba{2,}!` for achieving the same. This means a `b` followed by 2 or more `a`s, followed by a `!`.

In [21]:
for m in re.finditer(r'ba{2,}!', s):
    print(m)

<re.Match object; span=(16, 24), match='baaaaaa!'>
<re.Match object; span=(47, 52), match='baaa!'>
<re.Match object; span=(84, 96), match='baaaaaaaaaa!'>


### Match any character (except a newline)

The character `.` (dot or period) matches any character except a newline. This is the wildcard expression.

In [22]:
some_strings = ["begin", "began", "begun", "beg'n", "begone"]
for s in some_strings:
    m = re.search(r'beg.n', s)
    print(m)

<re.Match object; span=(0, 5), match='begin'>
<re.Match object; span=(0, 5), match='began'>
<re.Match object; span=(0, 5), match='begun'>
<re.Match object; span=(0, 5), match="beg'n">
<re.Match object; span=(0, 5), match='begon'>


### Match any string of characters

We use a period followed by Kleene * to match any string of characters. If you want to find all sentences that have at least two occurrences of the word 'ship', use the regex `ship.*ship`

In [23]:
s = "He worshipped on his ship."
m = re.search(r'ship.*ship', s)
print(m)

<re.Match object; span=(6, 25), match='shipped on his ship'>


#### `.*` won't match multiple lines due to carriage return in Windows

In Windows, the ENTER key produces two characters: a carriage return (`\r`) followed by a line feed (`\n`). You can use Notepad++ in Windows to display special characters. Click on View > Show Symbol > Show all characters.

In Linux, the ENTER key produces only one character: the line feed (`\n`).

Both of these are not matched by `.`

In [24]:
s = '''He worshipped on
his ship.'''

m = re.search(r'ship.*ship', s)
print(m)

None


#### How to make `.` match newline?

Use the flag `re.DOTALL`

In [25]:
s = '''He worshipped on
his ship.'''

m = re.search(r'ship.*ship', s, re.DOTALL)
print(m)

<re.Match object; span=(6, 25), match='shipped on\nhis ship'>


### Regexes are greedy by default

In [26]:
s = "He worshipped on his ship. He wanted to get that fellowship."

m = re.search(r'ship.*ship', s, re.DOTALL)
print(m)

<re.Match object; span=(6, 59), match='shipped on his ship. He wanted to get that fellow>


The `match` output above is confusing because the match is truncated! Notice that there's no closing quote so this is only the initial few characters of the match. We can get the full string that matched the regex using `m.group(0)`.

In [27]:
m.group(0)

'shipped on his ship. He wanted to get that fellowship'

#### Non-greedey regex with `.*?`

In [28]:
s = "He worshipped on his ship. He wanted to get that fellowship."

# Note that we have replaced .* with .*? below
m = re.search(r'ship.*?ship', s, re.DOTALL)
print(m)

<re.Match object; span=(6, 25), match='shipped on his ship'>


### Anchors

- `^`: beginning of a string or line
- `$`: end of string or line
- `\b`: word boundary
- `\B`: non word boundary

In [29]:
s = "The shop at the end of the street."
m = re.search(r'^The', s)
print(m)

<re.Match object; span=(0, 3), match='The'>


Note that the beginning anchor `^` itself does not match any character. It only enforces whatever follows it to be at the beginning.

In [30]:
s = '''They shopped at the end of the street. The street was empty.
Their enthusiasm got them going.
'''
for m in re.finditer(r'^The', s):
    print(m)

<re.Match object; span=(0, 3), match='The'>


By default, python re module anchors only at the beginning of string, not all new lines. To match all lines, we have to use the flag `re.MULTILINE`

In [31]:
for m in re.finditer(r'^The', s, re.MULTILINE):
    print(m)

<re.Match object; span=(0, 3), match='The'>
<re.Match object; span=(61, 64), match='The'>


In [32]:
# similarly, the `$` matches the end of lines.
# we want to match all occurrences of the string 'eat.' but only if it's at the end of line.

s = '''
Come on, this is where I want to eat.
The sheep are bleating nearby.
This is really neat.
I love this treat.
But please... no meat!
'''

In [33]:
for m in re.finditer(r'eat\.$', s, re.MULTILINE):
    print(m)

<re.Match object; span=(34, 38), match='eat.'>
<re.Match object; span=(86, 90), match='eat.'>
<re.Match object; span=(105, 109), match='eat.'>


Note that we had to "escape" the period with a backslash (`\`) above so it's not interpreted as the wildcard.

But what if you want to match only the word 'eat' and not occurrences of 'eat' within other words? Use `\b` to specify word boundaries on both sides of 'eat'.

In [34]:
for m in re.finditer(r'\beat\b', s):
    print(m)

<re.Match object; span=(34, 37), match='eat'>


And what if you want to match the word 'eat' only if it's part of a longer word? Use the non word boundary anchor `\B`.

In [35]:
for m in re.finditer(r'\Beat\B', s):
    print(m)

<re.Match object; span=(55, 58), match='eat'>


Now it matched the 'eat' in the word 'bleating'. But this is not complete. We also want to get the 'eat' in other words like 'neat', 'treat' and 'meat'. Use the disjunction operator pipe (`|`)

In [36]:
for m in re.finditer(r'\Beat|eat\B', s):
    print(m)

<re.Match object; span=(55, 58), match='eat'>
<re.Match object; span=(86, 89), match='eat'>
<re.Match object; span=(105, 108), match='eat'>
<re.Match object; span=(128, 131), match='eat'>


Now we have matched all occurrences of 'eat' that had a non-word character before it or after it.

### How to match both puppy and puppies?

In [37]:
s = '''Three little puppies. That puppy is especially cute!'''

# We can use disjunction on the full strings
for m in re.finditer(r'puppy|puppies', s):
    print(m)

<re.Match object; span=(13, 20), match='puppies'>
<re.Match object; span=(27, 32), match='puppy'>


In [38]:
# Or we can reuse only use disjunction for the end
for m in re.finditer(r'pupp(y|ies)', s):
    print(m)

<re.Match object; span=(13, 20), match='puppies'>
<re.Match object; span=(27, 32), match='puppy'>


## An exercise writing regex

Regexes, though handy, can become quite unweildy.

Let's write an RE to find cases of the English article 'the'. Consider the following string.

In [39]:
s = '''The North American Plate is a tectonic plate.
It extends eastward to the Mid-Atlantic Ridge and 
westward to the Chersky Range in eastern Siberia.
On the northerly boundary is a continuation of the 
Mid-Atlantic Ridge called the Gakkel Ridge. It moves
at 15mm - 25mm per year. The15mm speed was observed recently.'''

Here are the occurrences of the word 'the' in their contexts and the start indices of 'the' or 'The' in `s` within parantheses:

1. The North (0)
1. to the Mid (69)
1. to the Chersky (109)
1. On the northerly (150)
1. of the Mid (194)
1. called the Gakkel (225)
1. The15mm (277)

Note that we also want to count 'The15mm' (starting at index 277) even though it's a typo missing a space between 'The' and '15'.

In [40]:
# let us save the start indices, so we can use it for testing later

true_start_indices = [0, 69, 109, 150, 194, 225, 277]

#### Attempt 1

In [41]:
for m in re.finditer(r'the', s):
    print(m)

<re.Match object; span=(69, 72), match='the'>
<re.Match object; span=(109, 112), match='the'>
<re.Match object; span=(150, 153), match='the'>
<re.Match object; span=(157, 160), match='the'>
<re.Match object; span=(194, 197), match='the'>
<re.Match object; span=(225, 228), match='the'>


The regex missed two cases we wanted to capture: the ones starting at indices 0 and 277. In machine learning (ML) parlance, these instances are called "false negatives" (FN).

We can fix these by changing `the` to `[tT]he`.

The above regex also captures the 'the' within 'northerly' (start index 157). In ML parlance, this is a "false positive" (FP). We will deal with this next, but for the moment let's just try `[tT]he`.

#### Attempt 2

In [42]:
for m in re.finditer(r'[tT]he', s):
    print(m)

<re.Match object; span=(0, 3), match='The'>
<re.Match object; span=(69, 72), match='the'>
<re.Match object; span=(109, 112), match='the'>
<re.Match object; span=(150, 153), match='the'>
<re.Match object; span=(157, 160), match='the'>
<re.Match object; span=(194, 197), match='the'>
<re.Match object; span=(225, 228), match='the'>
<re.Match object; span=(277, 280), match='The'>


Now we got the 'The' at indices 0 and 277.

Next, let's fix the false positive at index 157 by requiring word boundaries around our regex.

#### Attempt 3

In [43]:
for m in re.finditer(r'\b[tT]he\b', s):
    print(m)

<re.Match object; span=(0, 3), match='The'>
<re.Match object; span=(69, 72), match='the'>
<re.Match object; span=(109, 112), match='the'>
<re.Match object; span=(150, 153), match='the'>
<re.Match object; span=(194, 197), match='the'>
<re.Match object; span=(225, 228), match='the'>


Now we are missing the 'The' starting at index 277 since it is followed immediately by the character '1', which is not a word boundary. So we can't use word boundary for our regex if we want to capture 'The15mm'. 

Let's modify the regex to allow for any character that is not `a-z` or `A-Z` to be around 'the'.

#### Attempt 4

In [44]:
for m in re.finditer(r'[^a-zA-Z][tT]he[^a-zA-Z]', s):
    print(m)

<re.Match object; span=(68, 73), match=' the '>
<re.Match object; span=(108, 113), match=' the '>
<re.Match object; span=(149, 154), match=' the '>
<re.Match object; span=(193, 198), match=' the '>
<re.Match object; span=(224, 229), match=' the '>
<re.Match object; span=(276, 281), match=' The1'>


Notice that now our match captures one character around the word as well.

But this is missing the 'The' at index 0, so we also need to allow for the beginning anchor before and ending anchor after.

#### Attempt 5

In [45]:
for m in re.finditer(r'(^|[^a-zA-Z])[tT]he([^a-zA-Z]|$)', s):
    print(m)

<re.Match object; span=(0, 4), match='The '>
<re.Match object; span=(68, 73), match=' the '>
<re.Match object; span=(108, 113), match=' the '>
<re.Match object; span=(149, 154), match=' the '>
<re.Match object; span=(193, 198), match=' the '>
<re.Match object; span=(224, 229), match=' the '>
<re.Match object; span=(276, 281), match=' The1'>


This regex now captures exactly what we wanted to capture.

Let us fix the code by getting the exact start indices. Anything with parantheses in the regex is available in `group`, so we can use `m.group(1)` to capture whatever is inside the first pair of parantheses, `m.group(2)` for the second pair, and so on. We can also use `m.span(1)` to get the start and end indices for the first group, etc.

Let's keep what we want to capture as the second group below.

In [46]:
for m in re.finditer(r'(^|[^a-zA-Z])([tT]he)([^a-zA-Z]|$)', s):
    print(m.group(2), m.span(2))

The (0, 3)
the (69, 72)
the (109, 112)
the (150, 153)
the (194, 197)
the (225, 228)
The (277, 280)


In [47]:
# Not let's keep the start indices in a variable and 
# compare against the true start indices.

regex_start_indices = []
for m in re.finditer(r'(^|[^a-zA-Z])([tT]he)([^a-zA-Z]|$)', s):
    regex_start_indices.append(int(m.span(2)[0]))

print(regex_start_indices)

[0, 69, 109, 150, 194, 225, 277]


In [48]:
assert(regex_start_indices == true_start_indices)

### Aliases for common sets of characters

In [56]:
# \d is for digit [0-9]
for m in re.finditer(r'\d', "Party of 5"):
    print(m)

<re.Match object; span=(9, 10), match='5'>


In [57]:
# \D is any non-digit i.e. [^0-9]
for m in re.finditer(r'\D', "747-800A"):
    print(m)

<re.Match object; span=(3, 4), match='-'>
<re.Match object; span=(7, 8), match='A'>


In [59]:
# \w is any alphanumeric/underscore [a-zA-Z0-9_]
for m in re.finditer(r'\w', "747-800A_B"):
    print(m)

<re.Match object; span=(0, 1), match='7'>
<re.Match object; span=(1, 2), match='4'>
<re.Match object; span=(2, 3), match='7'>
<re.Match object; span=(4, 5), match='8'>
<re.Match object; span=(5, 6), match='0'>
<re.Match object; span=(6, 7), match='0'>
<re.Match object; span=(7, 8), match='A'>
<re.Match object; span=(8, 9), match='_'>
<re.Match object; span=(9, 10), match='B'>


In [61]:
# \W is non-alphanumeric and not underscore [^\w]
for m in re.finditer(r'\W', "747-800A_B"):
    print(m)

<re.Match object; span=(3, 4), match='-'>


In [63]:
# \s is whitespace: space,tab,carriage return, line feed, form feed
# \s is the same as [ \r\t\n\f]
s = "hello there\tmy friend\nhow are you?"
print(f"s: {s}")
print('---')
for m in re.finditer(r'\s', s):
    print(m)

s: hello there	my friend
how are you?
---
<re.Match object; span=(5, 6), match=' '>
<re.Match object; span=(11, 12), match='\t'>
<re.Match object; span=(14, 15), match=' '>
<re.Match object; span=(21, 22), match='\n'>
<re.Match object; span=(25, 26), match=' '>
<re.Match object; span=(29, 30), match=' '>


In [66]:
# \S is non-whitespace [^\s]
for m in re.finditer(r'\S', "hi ok\tyes"):
    print(m)

<re.Match object; span=(0, 1), match='h'>
<re.Match object; span=(1, 2), match='i'>
<re.Match object; span=(3, 4), match='o'>
<re.Match object; span=(4, 5), match='k'>
<re.Match object; span=(6, 7), match='y'>
<re.Match object; span=(7, 8), match='e'>
<re.Match object; span=(8, 9), match='s'>


---

### Operators for counting

| Regex | Match |
|-------|-------|
| `*` | zero or more occurrences of the previous char or expression 
| `+` | one or more occurrences of the previous char or expression 
| `?` | zero or one occurrence of the previous char or expression 
| `{n}` | exactly $n$ occurrences of the previous char or expression 
| `{n,m}` | from $n$ to $m$ occurrences of the previous char or expression 
| `{n,}` | at least $n$ occurrences of the previous char or expression 
| `{,m}` | up to $m$ occurrences of the previous char or expression 

In [67]:
s = '''
Sheep 1 said "baaaa!""
Sheep 2 replied "baa!"
Sheep 3 said "baaaaaaaa!"
'''

# search for 'b' followed by 2 to 5 a's followed by !
for m in re.finditer(r'ba{2,5}!', s):
    print(m)

<re.Match object; span=(15, 21), match='baaaa!'>
<re.Match object; span=(41, 45), match='baa!'>


Escape these special characters when you want to literally match them:

`\*`
`\.`
`\?`

and
`\n` is for newline and 
`\t` is for tab.

### Substitutions

In python, substitutions are done with `re.sub`

In [72]:
# replace all occurrences of 'hello' with 'hi'
s = "hello there. ok, hello back."
s2 = re.sub(r'hello', 'hi', s)
print(s2)

hi there. ok, hi back.


### Back referencing

Problem: Put angle brackets around all integers in a string.

This requires back-referencing to the integer that's matched with the regex.

In [70]:
s = '''
This is 45th street. 
There are 320 homes here and about 2500 people live here.
'''

# the backslash 1 below refers back to the match;
# we use 1 since it's the first group within parantheses
s2 = re.sub(r'(\d+)', r'<\1>', s)
print(s2)


This is <45>th street. 
There are <320> homes here and about <2500> people live here.



Problem: find pattern "the Xer there were, the Xer they will be".
Here also we need backreference

In [73]:
my_strs = [
    'The lovelier they were, the lovelier they will be',
    'The sweeter they were, the lovelier they will be',
    'The calmer they were, the calmer they will be',
]   

In [75]:
my_pattern = re.compile(
    r'the (\w+)er they were, the \1er they will be', 
    re.IGNORECASE)

for s in my_strs:
    print(f"s: {s}")
    m = re.search(my_pattern, s)
    print(m)
    print('---')

s: The lovelier they were, the lovelier they will be
<re.Match object; span=(0, 49), match='The lovelier they were, the lovelier they will be>
---
s: The sweeter they were, the lovelier they will be
None
---
s: The calmer they were, the calmer they will be
<re.Match object; span=(0, 45), match='The calmer they were, the calmer they will be'>
---


### Non-capturing group

When you need to use parantheses for disjunction, but don't want to capture it in a group. Use `?:` right after the `(`.

In [76]:
# below, we don't want to capture the first disjunction in a group
my_pattern = re.compile(
    r'(?:some|a\s+few) (people|cats) like some \1',
    re.IGNORECASE
)

In [77]:
my_strs = [
    "A few people like some cats",
    "Some cats like some people",
    "Some cats like some cats",
    "A few cats like some cats",
]

for s in my_strs:
    print(s)
    for m in re.finditer(my_pattern, s):
        print(m)
    print('---')

A few people like some cats
---
Some cats like some people
---
Some cats like some cats
<re.Match object; span=(0, 24), match='Some cats like some cats'>
---
A few cats like some cats
<re.Match object; span=(0, 25), match='A few cats like some cats'>
---
